File: ai-js/src/tools/resource-download.tool.ts
import * as fs from 'fs/promises';
import * as path from 'path';
import axios from 'axios';
import { BaseTool, ToolParameter } from './tool.interface';

const DOWNLOAD_DIR = process.cwd() + '/tmp/download';

/**
 * 资源下载工具
 */
export class ResourceDownloadTool extends BaseTool {
  name = 'downloadResource';
  description = 'Download a resource from a given URL';
  parameters: ToolParameter[] = [
    {
      name: 'url',
      type: 'string',
      description: 'URL of the resource to download',
      required: true,
    },
    {
      name: 'fileName',
      type: 'string',
      description: 'Name of the file to save the downloaded resource',
      required: true,
    },
  ];

  async execute(args: Record<string, unknown>): Promise<string> {
    const url = args.url as string;
    const fileName = args.fileName as string;
    const filePath = path.join(DOWNLOAD_DIR, fileName);

    try {
      // 创建目录
      await fs.mkdir(DOWNLOAD_DIR, { recursive: true });

      // 下载文件
      const response = await axios.get(url, {
        responseType: 'arraybuffer',
        timeout: 30000,
      });

      await fs.writeFile(filePath, response.data);
      return `Resource downloaded successfully to: ${filePath}`;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      return `Error downloading resource: ${errorMessage}`;
    }
  }
}

File: src/main/java/com/github/tianchenghang/tools/ResourceDownloadTool.java
package com.github.tianchenghang.tools;

import cn.hutool.core.io.FileUtil;
import cn.hutool.http.HttpUtil;
import com.github.tianchenghang.constants.FileConstant;
import java.io.File;
import org.springframework.ai.tool.annotation.Tool;
import org.springframework.ai.tool.annotation.ToolParam;

public class ResourceDownloadTool {

  @Tool(description = "Download a resource from a given URL")
  public String downloadResource(
      @ToolParam(description = "Url of the resource to download") String url,
      @ToolParam(description = "Name of the file to save the downloaded resource")
          String fileName) {
    var fileDir = FileConstant.FILE_OUTPUT_DIR + "/download";
    var filepath = fileDir + "/" + fileName;
    try {
      // Make directory
      FileUtil.mkdir(fileDir);
      HttpUtil.downloadFile(url, new File(filepath));
      return "Resource downloaded successfully to: " + filepath;
    } catch (Exception e) {
      return "Error downloading resource: " + e.getMessage();
    }
  }
}
